File: sort/mergeSortLinkedList.java
public ListNode mergeSort(ListNode head) {
    //corner case
    if (head == null || head.next == null) {
        return head;
    }

    //find middle node and divide into two list
    ListNode mid = findMid(head);
    ListNode l = head;
    ListNode r = mid.next;
    mid.next = null;

    //go into the mergeSort
    ListNode h1 = mergeSort(l);
    ListNode h2 = mergeSort(r);
    ListNode result = merge(h1, h2);
    return result;
}
private ListNode findMid(ListNode head) {
    if (head == null || head.next == null) {
        return head;
    }
    ListNode slow = head;
    ListNode fast = head;
    //if 6 elemenet, 123456   if 5 elements, 12345
    //                 s f                     s f
    while (fast.next != null && fast.next.next != null) {
        fast = fast.next.next;
        slow = slow.next;
    }
    return slow;
}

public ListNode merge(ListNode one, ListNode two){
	ListNode dummy = new ListNode(0);
	ListNode cur = dummy;
	while(one != null && two != null){
		if(one.value <= two.value){
			cur.next = one;
			one = one.next;
		}else{
			cur.next = two;
			two = two.next;
		}
		cur = cur.next;
	}
	//lin the remaining possible nodes
	if(one != null)
		cur.next = one;
	else
		cur.next = two;
	return dummy.next;
}
//O(n) time,  O(n) for call stack, recursion not good as iterative
public ListNode merge(ListNode node1, ListNode node2){
	if(node1 == null)
		return node2;
	if(node2 == null)
		return node1;
	if(node1.val <= node2.val){
		node1.next = merge(node1.next, node2);
		return node1;
	}else{
		node2.next = merge(node1, node2.next);
		return node2;
	}
}

File: sort/mergeSortLinkedList.py
def mergesort(head):
    if not head or not head.next:
        return head

    # find middle and divide them into 2 parts
    midnode = findmid(head)
    left = head
    right = midnode.next
    midnode.next = None

    # recursion on each side
    lefthead = mergesort(left)
    righthead = mergesort(right)

    # merge
    finalhead = merge(lefthead, righthead)
    return finalhead

def findmid(head):
    if not head or not head.next:
        return head

    slow = head
    fast = head.next
    while not fast and not fast.next and
        slow = slow.next
        fast = fast.next.next
    return slow

# iterative merge
def merge(left, right):
    dummy = ListNode(0)
    cur = dummy
    while not left and not right:
        if left.value < right.value:
            cur.next = left
            left = left.next
        else:
            cur.next = right
            right = right.next
        cur = cur.next
    return dummy.next
    
# recursive merge
def merge(left, right):
    if not left:
        return right
    if not right:
        return left
    if left.value <= right.value:
        left.next = merge(left.next, right)
    else:
        right.next = merge(right.next, left)
